### ISSUE TITLE
Issue 6: Evaluate host-normalized features and shortcut mitigation on SCADANet

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/12-graph-based-cyber-physical-risk#27]
# Issue 6: Evaluate host-normalized features and shortcut mitigation on SCADANet

## Milestone: M3 — Leakage-Free Temporal and Shortcut Robustness

---

## Problem Statement

Implement and evaluate **host-relative feature normalization** for the SCADANet content features that were implicated in the previous false-positive analysis.

The previous work found that a large fraction of false positives were concentrated on two high-volume hosts. Threshold calibration reduced the false-positive rate, but did not address the underlying host-specific behavior.

This issue should test the deeper hypothesis.

---

## Research Hypothesis

A global feature distribution can make normal traffic from a busy/high-volume host look attack-like.

A host-relative baseline may better represent:

> Is this flow unusual for this host?

rather than:

> Is this flow unusual compared with all hosts?

---

### 1. Leakage-Safe Host-Normalization Method

Implement at least one leakage-safe host-normalization method.

#### Recommended approach — training-derived per-host z-score

For numeric feature `x` and source host `h`:

```text
z = (x - mean_train(h)) / std_train(h)
```

The host baseline must be estimated using **training data only**.

For hosts unseen during training, define and document a fallback, such as:

- global training statistics,
- destination-role baseline,
- neutral/default normalization.

Do not calculate host statistics using the test set.

#### Optional robust version

For heavily skewed traffic features, evaluate:

```text
(x - median_host) / IQR_host
```

if ordinary z-score behaves poorly.

Do not add many variants unless needed.

### 2. Features to test

At minimum apply host-relative treatment to relevant numeric content features, especially those previously examined such as:

- `frame_len`,
- `ip_len`,
- `Tcp_window_size`,
- other continuous packet/flow features.

For categorical features such as:

- `Protocol_TCP`
- `Tcp_flags_reset_Set`

do not z-score directly.

Instead consider host-relative frequency features if useful, for example:

```text
P(Tcp_flags_reset = Set | host, training period)
```

Only implement this if it can be done leakage-safely and cleanly.

### 3. Required Experimental Conditions

Use the same fixed evaluation split.

Compare:

#### H1 — Original model at default threshold

Reference.

#### H2 — Original model + leakage-safe threshold calibration

Existing mitigation.

#### H3 — Host-normalized features + default threshold

Tests the feature-level fix alone.

#### H4 — Host-normalized features + validation-selected threshold

Tests combined mitigation.

#### Optional H5 — Global normalization matched control

Useful if current preprocessing differs significantly.

### 4. Required Protocols

Run on:

1. SCADANet Track B.
2. Strict SCADANet temporal protocol from Issue #5.

The temporal experiment is important because host behavior may itself drift over time.

### 5. Required Host-Level Diagnostics

For each condition report:

- total false positives,
- normal-traffic FPR,
- false positives by source IP,
- false positives by destination IP,
- share of all false positives caused by top 1/top 2/top 5 hosts,
- recall by attack type,
- precision,
- F1,
- AUPRC.

Specifically track the previously problematic hosts.

### 6. Result Artifacts

Create:

```text
experiments/results/q1/host_normalization/
```

Suggested files:

```text
trackb_host_normalization_summary.csv
trackb_false_positives_by_host.csv
temporal_host_normalization_summary.csv
host_normalization_config.json
```

#### Table H1 — Mitigation comparison

| Condition | Precision | Recall | F1 | AUPRC | Normal FPR | Total FP |
|---|---:|---:|---:|---:|---:|---:|
| Original, threshold 0.50 | | | | | | |
| Original + calibrated threshold | | | | | | |
| Host-normalized | | | | | | |
| Host-normalized + calibrated threshold | | | | | | |

#### Table H2 — False-positive concentration by host

| Host | Original FP | Threshold-calibrated FP | Host-normalized FP | Combined FP |
|---|---:|---:|---:|---:|
| 192.168.119.140 | | | | |
| 192.168.119.141 | | | | |
| other hosts | | | | |

#### Table H3 — Attack recall before/after mitigation

| Attack type | Original recall | Host-normalized recall | Combined recall | Δ |
|---|---:|---:|---:|---:|

This guards against reducing false positives by simply missing attacks.

#### Figure H1 — False positives by host

Bar plot for the top false-positive-generating hosts under:

- original,
- calibrated,
- host-normalized,
- combined.

#### Figure H2 — Precision/FPR trade-off

Plot the four mitigation conditions in:

```text
x = FPR
y = recall or precision
```

A good mitigation moves toward low FPR without collapsing recall.

#### Figure H3 — Distribution before/after host normalization

For one or two representative continuous features:

- normal traffic from problematic hosts,
- normal traffic from other hosts,
- attack traffic.

Show before and after host normalization.

This figure should visually test the stated hypothesis.

---

### Strong support

Host normalization significantly reduces host-concentrated false positives while preserving attack recall.

Then the paper can claim:

> Host-relative preprocessing mitigates a specific shortcut identified through forensic and explainability analysis.

### Weak support

Host normalization has little effect.

Then retain threshold calibration as operational mitigation and report that the proposed root-cause correction was not sufficient.

### Trade-off

FPR falls but recall falls substantially.

Then the mitigation is not clearly preferable and should be described as a trade-off.

---

## Acceptance Criteria

- [ ] Host normalization uses training-derived statistics only.
- [ ] Unseen-host fallback is explicitly implemented.
- [ ] H1–H4 conditions are evaluated.
- [ ] Same split and same model family are used for fair comparison.
- [ ] FPR and attack recall are both reported.
- [ ] False-positive counts by host are saved.
- [ ] Top problematic hosts are analyzed explicitly.
- [ ] Table H1–H3 and Figure H1–H3 are generated automatically.
- [ ] The paper conclusion distinguishes threshold adjustment from root-cause feature mitigation.

---

## Do Not Do In This Issue

Do not:

- normalize using test-set host statistics,
- manually exclude problematic hosts,
- remove false positives from evaluation,
- optimize only F1 while ignoring FPR and attack recall.

The purpose is to test the host-shortcut hypothesis, not to make the metric prettier.
